src/Components/Components/src/Rendering/RendererSynchronizationContextDispatcher.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components.Rendering
+{
+    internal class RendererSynchronizationContextDispatcher : Dispatcher
+    {
+        private readonly RendererSynchronizationContext _context;
+
+        public RendererSynchronizationContextDispatcher()
+        {
+            _context = new RendererSynchronizationContext();
+        }
+
+        public override event UnhandledExceptionEventHandler UnhandledException
+        {
+            add
+            {
+                _context.UnhandledException += value;
+            }
+            remove
+            {
+                _context.UnhandledException -= value;
+            }
+        }
+
+        public override bool CheckAccess() => SynchronizationContext.Current == _context;
+
+        public override Task InvokeAsync(Action workItem)
+        {
+            if (CheckAccess())
+            {
+                workItem();
```

Comment:
What about sync exceptions?

```c#
try
{
    workItem();
    return Task.CompletedTask;
}
catch (Exception ex)
{
    return Task.FromException(ex);
}
```

Reply:
We don't generally try to protect you if you're doing error handling wrong. This kind of guidance was common during .NET 4.0 (pre-await) and doesn't make much sense these days. 

There are lots of places in the BCL that are async but will throw synchronous execptions.